fe/fe-core/src/main/java/com/starrocks/external/hive/HiveStatisticsProvider.java

```diff
@@ -0,0 +1,492 @@
+        OptionalDouble max =  columnStatistics.stream()
+                .map(HiveColumnStatistics::getMax)
+                .filter(value -> value != POSITIVE_INFINITY)
+                .mapToDouble(x -> x)
+                .max();
+        return max.isPresent() ? max.getAsDouble() : POSITIVE_INFINITY;
+    }
+
+    private double min(List<HiveColumnStatistics> columnStatistics) {
+        OptionalDouble min =  columnStatistics.stream()
+                .map(HiveColumnStatistics::getMin)
+                .filter(value -> value != NEGATIVE_INFINITY)
+                .mapToDouble(x -> x)
+                .min();
+        return min.isPresent() ? min.getAsDouble() : NEGATIVE_INFINITY;
+    }
+
+    private double getPerPartitionRowAvgNums(Collection<HivePartitionStatistics> statistics) {
+        if (statistics.isEmpty()) {
+            return -1;
+        }
+
+        long[] rowNums = statistics.stream()
+                .map(HivePartitionStatistics::getCommonStats)
+                .map(HiveCommonStats::getRowNums)
+                .filter(num -> num != -1)
+                .mapToLong(num -> num)
+                .toArray();
+
+        OptionalDouble avg = Arrays.stream(rowNums).average();
+        return avg.isPresent() ? avg.getAsDouble() : -1;
+    }
+
+    private double getValueFromLiteral(LiteralExpr literalExpr, Type type) {
+        switch (type.getPrimitiveType()) {
+            case TINYINT:
+            case SMALLINT:
+            case INT:
+            case BIGINT:
+                return literalExpr.getLongValue();
+            case LARGEINT:
+            case FLOAT:
+            case DOUBLE:
+            case DECIMALV2:
+            case DECIMAL32:
+            case DECIMAL64:
+            case DECIMAL128:
+                return literalExpr.getDoubleValue();
+            case DATE:
+            case DATETIME:
+                return (((DateLiteral) literalExpr).unixTimestamp(TimeZone.getDefault())) / 1000.0;
+            default:
+                return Double.NaN;
+        }
+    }
+
+    private double getLengthFromLiteral(LiteralExpr literalExpr, Type type) {
+        switch (type.getPrimitiveType()) {
+            case CHAR:
+            case VARCHAR:
+                return literalExpr.getStringValue().length();
+            default:
+                return type.getPrimitiveType().getSlotSize();
+        }
+    }
+
+    private static double normalizeFraction(double fraction) {
+        checkArgument(!isNaN(fraction), "fraction is NaN");
+        checkArgument(isFinite(fraction), "fraction must be finite");
+        if (fraction < 0) {
+            return 0;
+        }
+        if (fraction > 1) {
+            return 1;
+        }
+        return fraction;
+    }
+
+    public int getSamplePartitionSize(OptimizerContext session) {
+        return session.getSessionVariable().getHivePartitionStatsSampleSize();
+    }
+
+    static List<String> getPartitionsSample(List<String> partitions, int sampleSize) {
```

Comment:
add some comment to describe how to sample the partition

Reply:
done